File: Difficulty: Easy/Longest substring with distinct characters/longest-substring-with-distinct-characters.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
    int longestUniqueSubstr(string &s) {
        // code here
        int start=0, end=0, ans=0;
        unordered_map<char, int>mp;
        
        while(end<s.length()){
            mp[s[end]]++;
            while(mp[s[end]]==2){
                mp[s[start]]--;
                start++;
            }
            
            ans=max(ans, end-start+1);
            end++;
        }
        return ans;
    }
};


//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {

        Solution obj;
        string s;
        cin >> s;
        cout << obj.longestUniqueSubstr(s) << endl;
        cout << "~"
             << "\n";
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Longest substring with distinct characters/longest-substring-with-distinct-characters.js
//{ Driver Code Starts
// Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => { inputString += inputStdin; });

process.stdin.on("end", (_) => {
    inputString =
        inputString.trim().split("\n").map((string) => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let s = readLine().trim();
        let obj = new Solution();
        let res = obj.longestUniqueSubstr(s);
        console.log(res);

        console.log("~");
    }
}
// } Driver Code Ends


// User function Template for javascript

/**
 * @param {string} s
 * @return {number}
 */
class Solution {
    longestUniqueSubstr(s) {
        // code here
        let left =0
        let right =0
        let max =0
        let set = new Set()
        
        for(let i=0;i<s.length;i++){
            while(set.has(s[i])){
                set.delete(s[left])
                left++
            }
            set.add(s[i])
            if(set.size>max)max = set.size
        }
        return max
    }
}